### DISCUSSION TITLE
Proper way to attach CancellationToken to UniTask?

### DISCUSSION THREAD

**Discussion by o-tantk** [Cysharp/UniTask#585] [Q&A]
I'm doing something like this:
``` csharp
struct Foo {
    public Action OnBegin;
    public UniTask Yield;
    public Action OnEnd;
}

async UniTaskVoid BarAsync(Foo foo, CancellationToken token){
    foo.OnBegin?.Invoke();
    await foo.Yield; // <- I want to attach 'token' here.
    foo.OnEnd?.Invoke();
}
```

So... There is `AttachExternalCancellation` extension, but according to [Blog](https://neue.cc/2021/02/26_599.html) it seems better to avoid using.
> 名前的にも、使いたくない雰囲気を漂わせてる名前であるとおり、あんま使わないでねという意図が込められています。
> (MT) The name also has the intention of not using it too much, just as it gives off an atmosphere that you don't want to use.

Then, what is the proper way to attach cancellation token?
Or maybe should I change the logic?

**Comment by timcassell** [comment]
Instead of passing in an already-running task, you should pass in a delegate that takes the token as a parameter and returns the task. `Func<CancellationToken, UniTask>`

**Reply by o-tantk** [reply]
Is that the only way?
I wanted to pass UniTask.WaitUntil, UniTask.WaitForSeconds, etc. to the `Yield`. 😓
